### PR TITLE
Add ingest_pipeline data streams and dashboards to Elasticsearch package

### PR DESCRIPTION
**Pull request by joshdover**
## What does this PR do?

Part of https://github.com/elastic/kibana/issues/41936

This adds support to the existing Elasticsearch package for the new ingest_pipeline metricset added in Metricbeat in 8.7.0 in: https://github.com/elastic/beats/pull/34012. 

This started out as an httpjson input and eventually was moved to Metricbeat to improve it's availability to more customers. See [the main issue](https://github.com/elastic/beats/pull/34012) for more discussion.

Still TODO:
- [x] Determine a path forward on node-level granularity
	- Due to this limitation in Lens, we cannot create a line graph that shows the stats broken down by pipeline, summed across all nodes in the cluster: https://github.com/elastic/kibana/issues/146733
	- We can create tables that do this however
	- We either need to accept this limitation for now, wait until the Kibana issue is resolved, or remove the node-level granularity from the stats and sum the values across all nodes before sending to ES
	- **For now we're going to accept this limitation and keep the node-level granularity in the metrics.** I'll work with the Kibana team to prioritize the viz fix.
- [x] Add cluster name / ~ID~ to each document
	- cluster ID isn't available in node stats API so just using `name` for now
- [x] If we keep node-level granularity, add node name and role to each doc
- [x] Update the index pattern used to be able to reference `metrics-*,.monitoring-*,metricbeat-*` so it's generally useful for all collection methods
- [x] Add processor-level metrics, likely as a separate data stream
	- Challenges: this may create a very large volume of data that isn't likely to be used on a day-to-day basis. Need to determine a sampling or opt-in strategy for aiding optimization workflows separately from high-level monitoring
	- Decision: a new metricbeat configuration was added to sample process-level metrics. The default is 25%.
- [x] Implement collection and processing in metricbeat
	- [x] https://github.com/elastic/beats/pull/34012
	- [x] Update this integration to use underlying metricset
- [x] Update dashboard copy
	- [x] Reword "misbehaving processors" copy on pipeline detail dashboard to be more clear
	- [x] Add unit labels to axis where appropriate
	- [x] Double check consistent labels are used across visualizations, eg. "events per second" vs. "documents processed" etc.
- [x] Create and prioritize user stories and questions we want to be sure the dashboard can answer
	- [x] Add drilldowns & links into Ingest Pipeline editor and Stack Monitoring stats where appropriate
- [x] https://github.com/elastic/kibana/pull/149721
- [x] Determine how to handle the drilldowns to the pipeline editor when monitoring data is in a separate cluster
	- The "Edit pipeline" button doesn't work if the monitoring data is in a separate cluster from the one being monitored (which is typical).
	- See if we can provide a way to insert the Kibana URL of the cluster being monitored into the drilldown link to provide cross-linking
	- Otherwise, we may want to remove this for now.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Related issues

- Related to https://github.com/elastic/kibana/issues/41936

## Screenshots

image

image

### CODE REVIEW

File: packages/elasticsearch/data_stream/ingest/manifest.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+type: logs
+title: Elasticsearch ingest metrics
+release: experimental
+dataset: elasticsearch.stack_monitoring.ingest
```
Comment (miltonhultgren) [inline packages/elasticsearch/data_stream/ingest/manifest.yml:4]: We've named the other datasets with `stack_monitoring` to highlight that these are used to power the Stack Monitoring UI. 
For any new dataset we add it's best to omit this "tag".

File: packages/elasticsearch/data_stream/ingest/manifest.yml [hunk 2]
```diff
@@ -0,0 +1,12 @@
+type: logs
```
Comment (miltonhultgren) [inline packages/elasticsearch/data_stream/ingest/manifest.yml:1]: I'm a little confused, this is logs but isn't this metrics data coming out of the endpoints?

**Review by miltonhultgren — COMMENTED** [review]
This looks like what I imagine the future would be.

As we add new signals we'd expose them as new Integrations with their own datasets. It doesn't truly matter if they're using OpenTelemetry (although that will likely be the goal for anything that is metrics to replace the existing datasets).

The only thing we'd need to do is re-arrange the order the Integrations show up in and change the labels of the Stack Monitoring ones to highlight that these are legacy sources used to power the Stack Monitoring UI only.

Screenshot 2022-11-23 at 17 20 32

File: packages/elasticsearch/data_stream/ingest/manifest.yml (hunk 2 above)
Comment (joshdover) [reply packages/elasticsearch/data_stream/ingest/manifest.yml:1]: Yes it is. I haven't yet figured out how to configure the `httpjson` input to index as metrics via an integration, though I can get it to work with standalone Agent. To just make progress I omitted fixing this for now.

File: packages/elasticsearch/data_stream/ingest/agent/stream/stream.yml.hbs [hunk 3]
```diff
@@ -0,0 +1,24 @@
+config_version: "2"
+interval: 30s
+
+{{#if username}}
+auth.basic:
+  user: {{username}}
+  password: {{password}}
+{{/if}}
```
[the remark this answers is not in the document]
Comment (miltonhultgren) [reply packages/elasticsearch/data_stream/ingest/agent/stream/stream.yml.hbs:8]: Probably, I noticed we don't do that for the other data sets, we're tracking that here https://github.com/elastic/integrations/issues/4666
=

**Comment by philippkahr** [comment]
Love the design of this, I do have a few ideas for the dashboard:

We already have the cpu usage in % in the stack monitoring data, we could add this as a line view, normalised to 100% usage for all nodes?

image

I think a general chart that shows how much documents are ingested, vs how much are passing through an ingest pipeline? So I also am not mislead by seeing a drop.

**Comment by joshdover** [comment]
@philippkahr How's this look? I worry a bit that this makes the top chart a bit busy, but it may be worth it. We could average over all nodes instead of breaking down by node, but I don't think that's very useful.

image

**Comment by joshdover** [comment]
This has been updated with several new fields:
- `elasticsearch.node.name`
- `elasticsearch.node.roles`
- `elasticsearch.cluster.name`

We also now have drilldown links from the table views that will go either the Stack Monitoring page for node stats or the ingest pipeline editor.

**Comment by joshdover** [comment]
@jsoriano @hop-dev I'd like to have this dashboard reference a data view that covers both Agent data streams, Metricbeat indices, and `.monitoring-*` indices. Do we have any other integrations that do something similar (I couldn't find any)? Should I just add/export a data view / index pattern object directly to this package to accomplish this? Is that supported by Fleet?

**Comment by joshdover** [comment]
This is ready for review. It now uses the new `ingest_pipeline` metricset shipping in 8.7.0.

Note this bumps the minimum required version of the Stack to 8.7.0.

**Comment by joshdover** [comment]
Metricbeat code merged today. This should be testable after the next Beats and Agent snapshots are built.

**Comment by joshdover** [comment]
The current plan for this is to first release a GA version of the ES package with support for 8.5+ without this change and then merge this shortly after for the 8.7 release which will bump the minimum version for future versions to 8.7 due to the required Metricbeat changes.

Only downside of this is that testing the UI integration with the new dashboard won't be possible until this package is released.

**Comment by joshdover** [comment]
Currently failing right now with this error:
```
[elastic_agent][error] Unit state changed elasticsearch/metrics-default (STARTING->FAILED): [failed to reloading inputs: 1 error: Error creating runner from config: 1 error: metricset 'elasticsearch/ingest_pipeline' not found]
```

Traced through the code a bit and it's not obvious why this is happening. I suspect a stale build of Beats right now

**Comment by joshdover** [comment]
Confirmed that this is working on the latest 8.7.0-SNAPSHOT of Agent. This is ready for review @miltonhultgren and @klacabane

**Comment by miltonhultgren** [comment]
Looking at this now 👍🏼

File: packages/elasticsearch/changelog.yml [hunk 4]
```diff
@@ -1,4 +1,9 @@
 # newer versions go on top
+- version: "1.3.0-preview1"
+  changes:
+    - description: Add ingest pipeline monitoring metricset and dashboard
```
Comment (miltonhultgren) [inline packages/elasticsearch/changelog.yml:4]: Does it matter to call it dataset or metricset here?

**Review by miltonhultgren — APPROVED** [review]
This looks good to me! 

The one thing I'm not sure about is if there is a need to add more for the tests to work? I haven't look into Integration system testing yet.

File: packages/elasticsearch/changelog.yml (hunk 4 above)
Comment (joshdover) [reply packages/elasticsearch/changelog.yml:4]: Good call, let's use dataset.

**Comment by joshdover** [comment]
This is ready to be merged, however we will wait to merge it for a few more weeks. The reason being that we want some extra bake time on the recently GA'd ES package and want to be able to easily releases fixes to in Stack 8.5.0 before we introduce this PR which will only target 8.7.0+

**Comment by joshdover** [comment]
This is ready to be merged, however we will wait to merge it for a few more weeks. The reason being that we want some extra bake time on the recently GA'd ES package and want to be able to easily releases fixes to in Stack 8.5.0 before we introduce this PR which will only target 8.7.0+

**Comment by joshdover** [comment]
@klacabane @miltonhultgren How do you all feel about merging/releasing this tomorrow, as version 1.3.0 (GA). This would unblock testing of the new feature end-to-end on Cloud.

**Comment by klacabane** [comment]
That works. let's see if we can get https://github.com/elastic/integrations/pull/5353 merged as 1.3.0 first, then we can release this as 1.4.0 with the new kibana requirement

**Comment by joshdover** [comment]
Updated the version to 1.4.0

**Comment by joshdover** [comment]
Actually going to release this as 1.4.0-beta1 and then will promote to GA as a separate PR/release

**Pull request merged**